### PR TITLE
Increase validator robustness against mempool errors

### PR DESCRIPTION
**Pull request by styppo**
Allow validators to continue block production, even after failing to construct a block from the current mempool.

Log a detailed error message when block production fails.

fb224ca33f5819380247c8ed23139e69960dd500 adds an artificial panic to stop block production to help debug issue #2751. This should be reverted once that issue is fixed.

### CODE REVIEW

File: blockchain/src/block_production/mod.rs [hunk 1]
```diff
@@ -13,9 +13,43 @@ use nimiq_transaction::{
...
+    #[error("Failed to add to history")]
+    HistoryError,
+    #[error("Accounts are incomplete")]
+    AccountsIncomplete,
+}
+
+impl BlockProducerError {
+    fn accounts_error(
+        blockchain: &Blockchain,
+        error: AccountsError,
```
Comment (nibhar) [inline blockchain/src/block_production/mod.rs:33]:
```suggestion
    fn from_accounts_error(
        error: AccountsError,
        blockchain: &Blockchain,
```
I like this better, it is not a deal breaker though.

File: validator/src/micro.rs [hunk 2]
```diff
@@ -84,75 +84,103 @@ impl<TValidatorNetwork: ValidatorNetwork + 'static> NextProduceMicroBlockEvent<T
         let mut expected_next_ts;
 
         let return_value = loop {
+            // Wait for the expected timestamp to arrive before producing the block.
+            // We sleep at the beginning of the loop such that we can `continue` into the sleep.
+            // We can't sleep while holding the blockchain lock.
+            if delay > Duration::default() {
```
Comment (nibhar) [inline validator/src/micro.rs]:
```suggestion
            if delay.is_zero() {
```
I like `delay.is_zero()` better here, as it is in my opinion a more meaningful expression.

If yielding is not a problem the conditional could also be removed, sleeping unconditionally albeit for a zero duration. I did not check if that is problematic or not and I just assume that `sleep(Duration::ZERO).await;` will in fact yield.

File: validator/src/micro.rs [hunk 3]
```diff
@@ -84,75 +84,103 @@ impl<TValidatorNetwork: ValidatorNetwork + 'static> NextProduceMicroBlockEvent<T
...
+                        "Failed to construct micro block"
+                    );
+
+                    // Add an artificial panic here so that it's easier to catch any remaining
+                    // mempool bugs.
+                    // TODO Remove this.
+                    panic!("Failed to construct micro block");
+
+                    // Sleep a bit to allow the task to be cancelled externally if needed.
+                    delay = Duration::from_millis(50);
```
Comment (nibhar) [inline validator/src/micro.rs:143]: Mark this to allow unreachable code, for the time that the above panic is in place.

File: validator/src/micro.rs [hunk 4]
```diff
@@ -84,75 +84,103 @@ impl<TValidatorNetwork: ValidatorNetwork + 'static> NextProduceMicroBlockEvent<T
...
+                Err(error) => {
+                    error!(
+                        block_number = self.block_number,
+                        %error,
+                        "Failed to construct micro block"
+                    );
+
+                    // Add an artificial panic here so that it's easier to catch any remaining
+                    // mempool bugs.
+                    // TODO Remove this.
```
Comment (nibhar) [inline validator/src/micro.rs]: Please create an issue for this once this gets merged.

**Review by nibhar — APPROVED** [review]
Some minor comments, but it looks good to me.

**Review by Eligioo — APPROVED** [review]

**Pull request merged**